text/0000-upgrade-procedures

```diff
@@ -0,0 +1,56 @@
+
+### SDK changes
+
+Bugs fixes, security patches, and performance improvements are released as patch versions of the SDK. Support for new versions of the API are released as minor versions. New Plugin Interfaces and configuration options are also released as minor versions.
+
+Breaking changes to Plugin Interfaces are handled through deprecation. Instead of breaking a Plugin Interface, a new interface is created and the existing interface is marked as deprecated. Plugins which target the deprecated interface continue to work, and the SDK provides default implementations of any missing functionality. After one year, the deprecated Plugin Interface is removed as a major version release of the SDK.
+
+## Design requirements and explanations
+
+This approach to upgrading solves two critical design requirements, while minimizing the maintenance overhead associated with legacy code.
+
+* Callers of the API are never broken.
+* Users of the SDK can easily upgrade to the latest version.
+
+Indefinite support for existing Instrumentation is a critical feature of OpenTelemetry. Millions of lines of code are expected to be written against the API. This includes shared libraries which ship with integrated OpenTelemetry instrumentation. These libraries must be able to compose together into applications without OpenTelemetry creating a dependency conflict. While some Instrumentation will be updated to the latest version of the API, such as that provided by the OpenTelemetry project, other Instrumentation will never be updated.
+
+Consuming new Instrumentation will require users to upgrade to the latest version of the SDK. If it was not easy to perform this upgrade, the OpenTelemetry project would be forced to support older versions of the SDK, as well as older versions of the entire Instrumentation ecosystem. 
+
+This would be an enormous maintenance effort at best. But, as the OpenTelemetry project only controls a portion of that ecosystem, it is also unfeasible. OpenTelemetry cannot require that libraries with native instrumentation support multiple versions of the API. Ensuring that application owners and operators can upgrade to the latest version of the SDK resolves this issue.
+
+The primary blocker to upgrading the SDK is out of date Plugins. If a new version of the SDK were to break existing Plugin Interfaces, no user would be able to upgrade their SDK until the Plugins they depend on have been upgraded. Users could be caught between instrumentation they depend on requiring a version of the API which is not compatible with the version of the SDK which supports their Plugins.
+
+By following a deprecation pattern with Plugin Interfaces, we create a one year window in which the Plugin ecosystem can upgrade after the release of a new SDK. We believe this is sufficient time for any Plugin which is actively maintained to make an upgrade, and for defunct Plugins to be identified and replaced.
```

Comment:
I've been thinking a  lot about this here, and I want to through out an idea i've been noodling on.

1. Keep this 1-year plugin interface deprecation policy for most SDK plugin interfaces.
2. To alleviate concerns around 1-year timeline we provide a longer timeline on the Collector
  - We keep a 3-year compatibility guarantee on OTLP exporter to collector
  - We provide longer stability guarantees of "plugins API"s through the Collector

This is a non blocking comment, and just a thought I've had when reading this.  I think what you've written represents a balance of "what we can do" and "what users will want".   However, I know from "build plugin" space that 1-year can actually be "fast" for some old and highly used (but unchanged) plugins.   I'm not sure that'll be true in our space but want to call out the parallel.   Build plugins *likely* are more akin to instrumentation, but I still worry 1 year will be too short for the edge case scenarios we'll see people solve.

Reply:
I think that would be @tigrannajaryan/@bogdandrutu/collector-maintainers decision.

In general, I think we can support most deprecated plugin interfaces indefinitely, and the window is actually much wider. But I' prefer not to officially make such a long timeline a rule, as there may be edge cases where a Plugin Interface is being changed in order to support a change in the underlying framework architecture, and maintaining the old plugin interface is actually very burdensome.

Reply:
I would view this as a floor applied across the entire project.  If individual SIGs wanted to provide longer compatibility guarantees then I wouldn't see this as necessarily precluding them from doing so.

Reply:
I thought this document was exclusively about language APIs and SDKs. Do we need a separate doc, or at least a section, for the collector?

I think of the collector as having *only* plugin interfaces, no APIs. Also, since the collector is written in a single language (Go), we can be a lot more specific about the meaning of compatibility.

Reply:
@punya correct this document is just about the OpenTelemetry clients. In my opinion the Collector should have the same minimal guarantee (One year support for deprecated plugin interfaces), but I agree that there should be a stability and support document written which is specific to the Collector.